### PR TITLE
NFC: Move ReplacePushConstants pass to `Common/` for Metal

### PR DESCRIPTION
**Pull request by antiagainst**
We'll use the same pass to replace push constant access as normal uniform buffer access later.

Progress towards https://github.com/openxla/iree/issues/4370

### CODE REVIEW

**Comment by benvanik** [comment]
I think we want to use argument buffers and shouldn't need this pass. Is there a reason you aren't going to use them?

**Comment by benvanik** [comment]
(I'm guessing because you're routing through SPIR-V instead of generating msl?)

**Comment by antiagainst** [comment]
> I think we want to use argument buffers and shouldn't need this pass. Is there a reason you aren't going to use them?

Yes I'm [using](https://github.com/antiagainst/iree/blob/apple-metal-hal/experimental/metal/direct_command_buffer.m#L693) argument buffers for descriptor sets; that's pretty straightforward. Though I don't find good resources about push constant equivalences in Metal. It looks like https://developer.apple.com/documentation/metal/mtlargumentencoder/2915771-constantdataatindex is? From its doc: 

> index
>   This value corresponds to either the index ID of a Metal shading language declaration or the index field of a [MTLArgumentDescriptor](https://developer.apple.com/documentation/metal/mtlargumentdescriptor) object.

It suggests we can do either do `buffer(N)` style or `id(N)` style? (Although "within the argument buffer" is confusing to me.) The latter is encoding data into an argument buffer ("descriptor set"), where the former is more at the "pipeline" level so what we want for push constant. I'm not sure how to mix them though. Maybe call `constantDataAtIndex:` on the argument encoder before setting any active argument buffer?

> (I'm guessing because you're routing through SPIR-V instead of generating msl?)

Yeah, this is another problem. I tried to play with `constantDataAtIndex:`, but didn't figure out a way to let SPIRV-Cross generate the proper MSL code to mix the argument buffer style and old `buffer(N)` style and not let everything just crash.. 

But maybe it's due to my lack of experience. If you know pointers / resources, that'd be great! :)

File: compiler/src/iree/compiler/Codegen/Common/ReplacePushConstants.cpp [hunk 1]
```diff
@@ -21,8 +21,8 @@ namespace iree_compiler {
 namespace {
 
 // These must match what the runtime uses.
-#define IREE_HAL_WEBGPU_PARAMS_BIND_GROUP_INDEX 3
-#define IREE_HAL_WEBGPU_PARAMS_BINDING_INDEX 0
+#define IREE_HAL_PUSH_CONSTANT_SET_INDEX 3
+#define IREE_HAL_PUSH_CONSTANT_BINDING_INDEX 0
```
Comment (ScottTodd) [inline compiler/src/iree/compiler/Codegen/Common/ReplacePushConstants.cpp:25]: Some prior discussion about whether this pass should be shared: https://github.com/openxla/iree/pull/10224#discussion_r956462950

**Comment by dneto0** [comment]
> > I think we want to use argument buffers and shouldn't need this pass. Is there a reason you aren't going to use them?
> 
> Yes I'm [using](https://github.com/antiagainst/iree/blob/apple-metal-hal/experimental/metal/direct_command_buffer.m#L693) argument buffers for descriptor sets; that's pretty straightforward. Though I don't find good resources about push constant equivalences in Metal. It looks like https://developer.apple.com/documentation/metal/mtlargumentencoder/2915771-constantdataatindex is? From its doc:
> 

...
> 
> But maybe it's due to my lack of experience. If you know pointers / resources, that'd be great! :)

I think setBytes is what you want.
See https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443159-setbytes?language=objc


> ## Discussion
> Using this method is equivalent to creating a new [MTLBuffer](https://developer.apple.com/documentation/metal/mtlbuffer?language=objc) object from the specified data and then binding it to the compute shader with the [setBuffer:offset:atIndex:](https://developer.apple.com/documentation/metal/mtlcomputecommandencoder/1443126-setbuffer?language=objc) method. However, this method avoids the overhead of creating a buffer to store your data; instead, Metal manages the data.

> Use this method for single-use data smaller than 4 KB. Create a [MTLBuffer](https://developer.apple.com/documentation/metal/mtlbuffer?language=objc) object if your data exceeds 4 KB in length or persists for multiple uses.

**Comment by antiagainst** [comment]
@dneto0: yup indeed! thanks for the pointers!

After some searching I found out the way to teach SPIRV-Cross to emit the MSL I'd expect: https://github.com/KhronosGroup/SPIRV-Cross/issues/2098#issuecomment-1406377858. So this is not needed any more. 

Thanks everyone for the help! :)

**Pull request closed without merging**